Validator fails when field value is null.

Given the following schema
```json
{
    "namespace": "org.srem",
    "type": "record",
    "name": "test_case",
    "doc": "test_schema",
    "fields": [
        {
            "name": "event_id",
            "type": "string",
            "doc": "ID of the event in UUID4 format"
        },
        {
            "name": "payload",
            "doc": "Variable payload depending on client",
            "type": [
                "null",
                {
                    "type": "array",
                    "items": "string"
                },
                {
                    "type": "map",
                    "values": [
                        "null",
                        "string",
                        "int"
		]
                    
                }
            ]
        },
        {
            "name": "__metadata",
            "type": [
                "null",
                {
                    "type": "map",
                    "values": "string"
                }
            ],
            "default": null,
            "doc": "Field used in Hermes internals to propagate metadata like hermes-id"
        }
    ]
}
```
and the following json data
```json
{
    "event_id": "e130a7ca802a494ebf1997b1e340df71",
    "payload": {
        "foo": null
    }
}
```

the validator fails with an NPE:
```java
2017-01-12 17:06:09.839 DEBUG t.a.s.j.v.schema.avro.AvroValidator - Converting JSON to AVRO
2017-01-12 17:06:09.883 ERROR t.a.s.j.validator.ValidatorRunner - Document could not be validated
tech.allegro.schema.json2avro.validator.schema.ValidatorException: Error occurred when validating the document
	at tech.allegro.schema.json2avro.validator.schema.avro.AvroValidator.convertJsonToAvro(AvroValidator.java:74) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.validator.schema.avro.AvroValidator.validate(AvroValidator.java:46) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.validator.ValidatorRunner.run(ValidatorRunner.java:27) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.validator.ValidatorRunner.main(ValidatorRunner.java:38) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
Caused by: java.lang.NullPointerException: null
	at java.util.HashMap.merge(HashMap.java:1224) ~[na:1.8.0_111]
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320) ~[na:1.8.0_111]
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[na:1.8.0_111]
	at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[na:1.8.0_111]
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[na:1.8.0_111]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_111]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_111]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_111]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_111]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_111]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.readMap(JsonGenericRecordReader.java:96) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.lambda$read$4(JsonGenericRecordReader.java:70) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.onValidType(JsonGenericRecordReader.java:132) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.read(JsonGenericRecordReader.java:70) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.readUnion(JsonGenericRecordReader.java:103) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.read(JsonGenericRecordReader.java:71) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.lambda$null$0(JsonGenericRecordReader.java:55) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at java.util.Optional.ifPresent(Optional.java:159) ~[na:1.8.0_111]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.lambda$readRecord$1(JsonGenericRecordReader.java:55) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at java.util.LinkedHashMap$LinkedEntrySet.forEach(LinkedHashMap.java:671) ~[na:1.8.0_111]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.readRecord(JsonGenericRecordReader.java:53) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.read(JsonGenericRecordReader.java:45) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonGenericRecordReader.read(JsonGenericRecordReader.java:36) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonAvroConverter.convertToGenericDataRecord(JsonAvroConverter.java:47) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.converter.JsonAvroConverter.convertToAvro(JsonAvroConverter.java:38) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	at tech.allegro.schema.json2avro.validator.schema.avro.AvroValidator.convertJsonToAvro(AvroValidator.java:70) ~[json2avro-validator-0.2.5-SNAPSHOT.jar:na]
	... 3 common frames omitted
```

This translates fine to avro using python avro 1.8.1, so the validator shouldn't fail on this.

--- a/converter/src/main/java/tech/allegro/schema/json2avro/converter/JsonGenericRecordReader.java
+++ b/converter/src/main/java/tech/allegro/schema/json2avro/converter/JsonGenericRecordReader.java
@@ -10,6 +10,7 @@ import org.codehaus.jackson.map.ObjectMapper;
 import java.io.IOException;
 import java.util.ArrayDeque;
 import java.util.Deque;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Function;
@@ -91,9 +92,9 @@ public class JsonGenericRecordReader {
     }
 
     private Map<String, Object> readMap(Schema.Field field, Schema schema, Map<String, Object> map, Deque<String> path) {
-        return map.entrySet()
-                .stream()
-                .collect(toMap(Map.Entry::getKey, entry -> read(field, schema.getValueType(), entry.getValue(), path, false)));
+        Map<String, Object> result = new HashMap<>(map.size());
+        map.forEach((k, v) -> result.put(k, read(field, schema.getValueType(), v, path, false)));
+        return result;
     }
 
     private Object readUnion(Schema.Field field, Schema schema, Object value, Deque<String> path) {

--- a/converter/src/test/groovy/tech/allegro/schema/json2avro/converter/JsonAvroConverterSpec.groovy
+++ b/converter/src/test/groovy/tech/allegro/schema/json2avro/converter/JsonAvroConverterSpec.groovy
@@ -779,6 +779,50 @@ class JsonAvroConverterSpec extends Specification {
         exception.cause.message  ==~ /.*enum type and be one of A, B, C.*/
     }
 
+    def 'should accept null when value can be of any nullable array/map type'() {
+        given:
+        def schema = '''
+            {
+                "type": "record",
+                "name": "testSchema",
+                "fields": [
+                    {
+                        "name": "payload",
+                        "type": [
+                            "null",
+                            {
+                                "type": "array",
+                                "items": "string"
+                            },
+                            {
+                                "type": "map",
+                                "values": [
+                                    "null",
+                                    "string",
+                                    "int"
+                                ]
+                             }
+                        ]
+                    }
+                ]
+            }
+        '''
+
+        def json = '''
+        {
+            "payload": {
+                "foo": null
+            }
+        }
+        '''
+
+        when:
+        byte[] avro = converter.convertToAvro(json.bytes, schema)
+
+        then:
+        !toMap(converter.convertToJson(avro, schema)).payload.foo
+    }
+
     def toMap(String json) {
         slurper.parseText(json)
     }

